closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -601,6 +603,7 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
     if ((NULL == OSSL_error_string) ||
         (NULL == OSSL_error_string_n) ||
         (NULL == OSSL_get_error) ||
+        (NULL == OSSL_md5) ||
```

Comment:
Support for md5 is optional in OpenSSL, so this should probably not be a fatal error that disables use of all other algorithms.

Reply:
I'm not sure what you mean by optional. Are there versions of OpenSSL that don't have `MD5`?

Reply:
I was referring to the configuration flag `OPENSSL_NO_MD5` used in [`crypto/evp/c_alld.c`](https://github.com/openssl/openssl/blob/openssl-3.1/crypto/evp/c_alld.c#L22).
We shouldn't sacrifice all other uses of OpenSSL just because the library that's available was configured without md5 support.

Reply:
According to [this](https://manpages.opensuse.org/Tumbleweed/libressl-devel-doc/EVP_add_digest.3ssl.en.html), the `EVP_add_digest()` call adds `MD5` to a global lookup table so that it can be retrieved with `EVP_get_cipherbyname()`.

But in the code I get it directly through `EVP_md5()`. I am not sure the library loading will fail, even if `OPENSSL_NO_MD5` is set. What do you think?

Reply:
We use `EVP_md5()` (https://www.openssl.org/docs/man3.0/man3/EVP_md5.html) which is not deprecated. `MD5()` (https://www.openssl.org/docs/man3.0/man3/MD5.html) is the one that's actually marked for deprecation.

You are right, however, in that someone could build with `no-md5` and force OpenSSL to not export those, so I can definitely add a check for that.

Only problem is that same thing could be done for any of `aes`, `aria`, `bf`, `camellia`, `cast`, `des`, `dh`, `dsa`, `ec`, `hmac`, `idea`, `md2`, `md5`, `mdc2`, `rc2`, `rc4`, `rc5`, `ripemd`, `seed`, `sha`, `sm2`, `sm3`, `sm4`. Do you think we should make a note to add check for each of them in the future if possible rather than failing to load?

Reply:
I added an additional method to check whether MD5 is available instead of failing to load OpenSSL altogether.

Reply:
I don't think we're using any of the other algorithms listed above (yet?), so I don't think we need to do anything right now. If and when we add support for other algorithms, we should consider whether they may become deprecated and thus not available, or similarly, new algorithms that might not be present in older libraries.

Reply:
We do use `SHA` and `EC`. And it's not about deprecation, but rather building OpenSSL with `-no-<cipher>` like you mentioned for `MD5`.